### DISCUSSION TITLE
Is it possible to save as csv instead of tsv by default?

### DISCUSSION THREAD

**Discussion by skeept** [saulpw/visidata#1022] [Q&A] [answered]
I have a file that ends with extension `.dat` which I open with

```
vd --filetype csv --header 0 output.dat
```

if I make any change and try to save it is saved as a tsv file with the message 
`save for dat unavailable, using tsv` but this changes the file as it was a csv file before. Since I did specify from the command line that this is a csv file I would expect it to save it as one as  well...
Is it possible to change some setting to have it saved with csv format?
I know I change the output name to .csv, but then I need the extra step to rename the file...

Also, related to this, would it be possible to change a setting that whenever a file ends with `.dat` assume `--filetype csv` and `--header 0` from visidata directly?

thank you.

**Comment by anjakefala** [comment, marked as answer]
Hi @skeept!

>  Is it possible to change some setting to have it saved with csv format?

Yes! The option to change the default save filetype is `--save-filetype`. It is different from the specified loading filetype (`--filetype`). So you would do `--save-filetype csv`. There is both a `--filetype` and `--save-filetype`, so that a user can specify the filetype for conversion.


> Also, related to this, would it be possible to change a setting that whenever a file ends with .dat assume --filetype csv and --header 0 from visidata directly?

What a fun kata! If you put this in your `~/.visidatarc`, it should do the trick. =) 

```
@VisiData.api
def open_dat(vd, p):
    return DatSheet(p.name, source=p)
    
class DatSheet(CsvSheet):
        pass

DatSheet.class_options.header=0
vd.save_dat = vd.save_csv
```

Though, I am not sure how this will behave with the fact that we do have an `open_dat` within the pandas loader. (I.e. if you used `--filetype pandas`, the pandas loader would know how to load your `.dat` file).

**Reply by skeept** [reply]
@anjakefala thank you for your answer. That's exactly what I was looking for. I glanced at the man page several times but didn't realize the `--save-filetype` option was there.

The code you provided seems to work great too except that I need to remove the 
```
@VisiData.api
```
decorator part. If take that out it works fine but if I leave that there it fails with an error:

```
Traceback (most recent call last):
  File "/conda/bin/vd", line 6, in <module>
    visidata.main.vd_cli()
  File "/conda/lib/python3.8/site-packages/visidata/main.py", line 301, in vd_cli
    rc = main_vd()
  File "/conda/lib/python3.8/site-packages/visidata/main.py", line 215, in main_vd
    vs = vd.openSource(p, create=True, **opts) or vd.fail(f'could not open {p}')
  File "/conda/lib/python3.8/site-packages/visidata/_open.py", line 90, in openSource
    vs = vd.openPath(Path(p), filetype=filetype, create=create)  # convert to Path and recurse
  File "/conda/lib/python3.8/site-packages/visidata/_open.py", line 73, in openPath
    return openfunc(p)
TypeError: open_dat() takes 1 positional argument but 2 were given
```

**Reply by saulpw** [reply]
Hey @skeept, nice job diagnosing/fixing that.  The "global" open_ functions are all-but-deprecated, so a better fix would be to keep the `@VisiData.api` and add `vd` as the first argument: `def open_dat(vd, p):`

**Reply by skeept** [reply]
Thank you, that worked, I will use latest recommendation.

**Comment by skeept** [comment]
just a note, in recent versions of visidata the part 

```
class DatSheet(CsvSheet):
        pass
```

doesn't work anymore. It needs to be
```
class DatSheet(visidata.loaders.csv.CsvSheet):
        pass
```

**Reply by anjakefala** [reply]
Thanks for letting us know! `CsvSheet` is back in globals on `develop`, thanks to @geekscrapy.
